File: py/arolla/types/qtype/optional_qtype.py
# Copyright 2024 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""(Private) Optional qtypes and corresponding factory functions.

Please avoid using this module directly. Use arolla (preferrably) or
arolla.types.types instead.
"""

from arolla.abc import abc as arolla_abc
from arolla.types.qtype import casting
from arolla.types.qtype import clib
from arolla.types.qtype import scalar_qtype

# Returns OPTIONAL_BOOLEAN qvalue.
optional_boolean = clib.optional_boolean

# Returns OPTIONAL_BYTES qvalue.
optional_bytes = clib.optional_bytes

# Returns OPTIONAL_FLOAT32 qvalue.
optional_float32 = clib.optional_float32

# Returns OPTIONAL_FLOAT64 qvalue.
optional_float64 = clib.optional_float64

# Returns OPTIONAL_INT32 qvalue.
optional_int32 = clib.optional_int32

# Returns OPTIONAL_INT64 qvalue.
optional_int64 = clib.optional_int64

# Returns OPTIONAL_TEXT qvalue.
optional_text = clib.optional_text

# Returns OPTIONAL_UINT64 qvalue.
optional_uint64 = clib.optional_uint64

# Returns OPTIONAL_UNIT qvalue.
optional_unit = clib.optional_unit

# Returns OPTIONAL_WEAK_FLOAT qvalue.
optional_weak_float = clib.optional_weak_float

# LINT.IfChange
OPTIONAL_BOOLEAN = optional_boolean(None).qtype
OPTIONAL_BYTES = optional_bytes(None).qtype
OPTIONAL_FLOAT32 = optional_float32(None).qtype
OPTIONAL_FLOAT64 = optional_float64(None).qtype
OPTIONAL_INT32 = optional_int32(None).qtype
OPTIONAL_INT64 = optional_int64(None).qtype
OPTIONAL_TEXT = optional_text(None).qtype
OPTIONAL_UINT64 = optional_uint64(None).qtype
OPTIONAL_UNIT = optional_unit(None).qtype
OPTIONAL_WEAK_FLOAT = optional_weak_float(None).qtype

OPTIONAL_SCALAR_SHAPE = clib.OPTIONAL_SCALAR_SHAPE
# LINT.ThenChange(//depot/py/arolla/rl.py)


OPTIONAL_QTYPES = (
    OPTIONAL_UNIT,
    OPTIONAL_BOOLEAN,
    OPTIONAL_BYTES,
    OPTIONAL_TEXT,
    OPTIONAL_INT32,
    OPTIONAL_INT64,
    OPTIONAL_UINT64,
    OPTIONAL_FLOAT32,
    OPTIONAL_FLOAT64,
    OPTIONAL_WEAK_FLOAT,
)


def is_optional_qtype(qtype: arolla_abc.QType) -> bool:
  """Indicates whether the given qtype is optional."""
  return casting.get_shape_qtype_or_nothing(qtype) == OPTIONAL_SCALAR_SHAPE


def make_optional_qtype(qtype: arolla_abc.QType) -> arolla_abc.QType:
  """Returns the optional qtype for a given value qtype."""
  if not (scalar_qtype.is_scalar_qtype(qtype) or is_optional_qtype(qtype)):
    raise casting.QTypeError(f'expected a scalar qtype, got {qtype=}')
  return casting.broadcast_qtype([OPTIONAL_UNIT], qtype)


def missing_unit() -> arolla_abc.AnyQValue:
  """Returns a missing OPTIONAL_UNIT."""
  return optional_unit(None)


def present_unit() -> arolla_abc.AnyQValue:
  """Returns a present OPTIONAL_UNIT."""
  return optional_unit(True)


# Aliases
missing = missing_unit
present = present_unit

File: arolla/qtype/optional_qtype.cc
// Copyright 2024 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
#include "arolla/qtype/optional_qtype.h"

#include "absl/base/no_destructor.h"
#include "absl/base/thread_annotations.h"
#include "absl/container/flat_hash_map.h"
#include "absl/log/check.h"
#include "absl/log/log.h"
#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"
#include "absl/synchronization/mutex.h"
#include "arolla/memory/frame.h"
#include "arolla/qtype/base_types.h"
#include "arolla/qtype/qtype.h"
#include "arolla/qtype/qtype_traits.h"
#include "arolla/qtype/typed_value.h"
#include "arolla/util/unit.h"

namespace arolla {
namespace {

class OptionalQTypeMaps {
 public:
  // Register a qtype, optional_qtype pair.
  void Register(QTypePtr qtype, QTypePtr optional_qtype) {
    absl::MutexLock l(&lock_);
    to_optional_[qtype] = optional_qtype;
    to_optional_[optional_qtype] = optional_qtype;
  }

  // Given a qtype, return corresponding optional qtype if it exists.
  absl::StatusOr<QTypePtr> ToOptionalQType(QTypePtr qtype) {
    absl::ReaderMutexLock l(&lock_);
    auto iter = to_optional_.find(qtype);
    if (iter == to_optional_.end()) {
      return absl::InvalidArgumentError(
          absl::StrCat("no optional qtype for ", qtype->name()));
    }
    return iter->second;
  }

  // Check whether the given QType is an optional QType.
  bool IsOptionalQType(QTypePtr qtype) {
    absl::ReaderMutexLock l(&lock_);
    auto iter = to_optional_.find(qtype);
    return iter != to_optional_.end() && iter->second == qtype;
  }

 private:
  absl::Mutex lock_;
  absl::flat_hash_map<QTypePtr, QTypePtr> to_optional_ ABSL_GUARDED_BY(lock_);
};

OptionalQTypeMaps* GetOptionalQTypeMaps() {
  static absl::NoDestructor<OptionalQTypeMaps> instance;
  return instance.get();
}

}  // namespace

void RegisterOptionalQType(QTypePtr optional_qtype) {
  const auto* value_qtype = optional_qtype->value_qtype();
  DCHECK(value_qtype != nullptr);
  const auto& sub_slots = optional_qtype->type_fields();
  DCHECK_GT(sub_slots.size(), 0);
  DCHECK(sub_slots[0].GetType()->type_info() == typeid(bool));
  DCHECK_EQ(sub_slots[0].byte_offset(), 0);
  if (sub_slots.size() == 1) {
    DCHECK_EQ(value_qtype, GetQType<Unit>());
  } else if (sub_slots.size() == 2) {
    DCHECK_EQ(sub_slots[1].GetType(), value_qtype);
  } else {
    LOG(FATAL) << "Unexpected number of subslots in optional: "
                << sub_slots.size();
  }
  GetOptionalQTypeMaps()->Register(value_qtype, optional_qtype);
}

absl::StatusOr<QTypePtr> ToOptionalQType(QTypePtr qtype) {
  return GetOptionalQTypeMaps()->ToOptionalQType(qtype);
}

const QType* /*nullable*/ DecayOptionalQType(const QType* /*nullable*/ qtype) {
  return IsOptionalQType(qtype) ? qtype->value_qtype() : qtype;
}

bool IsOptionalQType(const QType* /*nullable*/ qtype) {
  // Use the properties we verified earlier to discard the non-optional types
  // before locking the mutex.
  return (qtype != nullptr && qtype->value_qtype() != nullptr &&
          !qtype->type_fields().empty() &&
          GetOptionalQTypeMaps()->IsOptionalQType(qtype));
}

absl::StatusOr<FrameLayout::Slot<bool>> GetPresenceSubslotFromOptional(
    TypedSlot slot) {
  if (!IsOptionalQType(slot.GetType())) {
    return absl::InvalidArgumentError(
        absl::StrCat("'", slot.GetType()->name(), "' is not optional qtype."));
  }
  if (slot.SubSlotCount() == 0) {
    return absl::InternalError("optional value has no subslots.");
  }
  return slot.SubSlot(0).ToSlot<bool>();
}

absl::StatusOr<TypedSlot> GetValueSubslotFromOptional(TypedSlot slot) {
  if (!IsOptionalQType(slot.GetType())) {
    return absl::InvalidArgumentError(
        absl::StrCat("'", slot.GetType()->name(), "' is not optional qtype."));
  }
  if (slot.SubSlotCount() != 2) {
    return absl::InvalidArgumentError(absl::StrCat(
        "'", slot.GetType()->name(), "' does not have a value subslot."));
  }
  return slot.SubSlot(1);
}

absl::StatusOr<TypedValue> CreateMissingValue(QTypePtr optional_qtype) {
  if (!IsOptionalQType(optional_qtype)) {
    return absl::InvalidArgumentError(absl::StrFormat(
        "cannot create a missing value for non-optional qtype `%s`",
        optional_qtype->name()));
  }
  return TypedValue::UnsafeFromTypeDefaultConstructed(optional_qtype);
}

}  // namespace arolla
